### PR TITLE
feat(nns): Store `wasm_metadata` in SNS-W's stable memory (attempt #2)

### PR DESCRIPTION
**Pull request by aterga**
This PR is the first step towards enabling the `SnsW.get_wasm_metadata` endpoint in production. For that, we need a way to store the metadata in the stable memory, as recomputing it each time is expensive (because canister WASMs stored in SNS-W are often compressed).

Unlike the [first attempt](https://github.com/dfinity/ic/pull/956), this PR makes it harder to represent disallowed states in the type system, e.g., there cannot metadata that does not belong to any WASM stored in SNS-W.

### CODE REVIEW

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 1]
```diff
@@ -422,12 +423,30 @@ where
...
+                println!("err = {}, wasm = `{:?}`", err, wasm);
+                return AddWasmResponse {
+                    result: Some(add_wasm_response::Result::Error(SnsWasmError {
+                        message: format!("Cannot read metadata sections from WASM: {}", err),
+                    })),
+                };
+            }
+        };
+
+        let metadata = metadata.into_iter().map(MetadataSectionPb::from).collect();
```
Comment (aterga) [inline rs/nns/sns-wasm/src/sns_wasm.rs:448]: At this point, we need to decide what to do if the metadata is too big.

Any suggestions how to best restrict it?

If we don't enforce any limits here, in the worst case, storing the metadata in SNS-W would increase the memory consumption by almost 2X.

File: bazel/external_crates.bzl [hunk 2]
```diff
@@ -735,7 +735,7 @@ def external_crates_repository(name, cargo_lockfile, lockfile, sanitizers_enable
                 version = "^0.2.155",
             ),
             "libflate": crate.spec(
-                version = "^1.1.2",
+                version = "^2.0.0",
```
Comment (aterga) [inline bazel/external_crates.bzl:738]: Why I'm bumping this dependency version in this PR: 

This version 2.0.0 corresponds to what is currently used in ic-wasm 0.8.0

**Comment by aterga** [comment]
TODO: test the migration code

File: Cargo.Bazel.Fuzzing.json.lock [hunk 3]
```diff
@@ -13243,6 +13243,61 @@
       ],
       "license_file": "LICENSE"
     },
+    "core2 0.4.0": {
```
Comment (max-dfinity) [inline Cargo.Bazel.Fuzzing.json.lock:13246]: did you mean to cause so many updates to cargo files?

File: rs/nervous_system/common/test_utils/src/wasm_helpers.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        metadata::Kind::Private
+    };
+
+    let mut wasm_module = utils::parse_wasm(wasm, false).expect("Cannot parse WASM");
+    metadata::add_metadata(&mut wasm_module, kind, name, contents);
+    wasm_module.emit_wasm()
+}
+
+// Gzips a wasm, returning the hash of its compressed representation.
+pub fn gzip_wasm(wasm: &[u8]) -> Vec<u8> {
```
Comment (max-dfinity) [inline rs/nervous_system/common/test_utils/src/wasm_helpers.rs:29]: I think there's a function somewhere called something like "ensure_gzipped" so that you can pass it bytes that are gzipped or not, and either zip it or leave it alone depending.

File: rs/nns/sns-wasm/canister/canister.rs [hunk 5]
```diff
@@ -334,6 +335,12 @@ fn canister_post_upgrade() {
     println!("{}Completed post upgrade", LOG_PREFIX);
 }
 
+// TODO[NNS1-2997]: Remove this function once the migration is completed.
+#[heartbeat]
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/canister/canister.rs:339]: Let's use the dfn_core version of heartbeat until we migrate the canister to ic_cdk?

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 6]
```diff
@@ -2134,10 +2205,31 @@ mod test {
         state
     }
 
-    /// Add some placeholder wasms with different values so we can test
-    /// that each value is installed into the correct spot
-    /// group_number tells you which group of wasms you are adding so that they
-    /// will have different content and therefore different hashes
+    // Adds section "icp:[public|private] $name$contents" to `wasm`, returning `wasm`'s new hash.
+    fn annotate_wasm_with_metadata(
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:2209]: This is confusing, b/c it returns a hash instead of a wasm that's annotated.  Could we fix the function name?

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 7]
```diff
@@ -4917,40 +5009,11 @@ mod test {
...
-            let mut wasm_module = utils::parse_wasm(&wasm.wasm, false).unwrap();
-            metadata::add_metadata(&mut wasm_module, kind, name, contents);
-
-            wasm.wasm = wasm_module.emit_wasm();
-
-            Sha256::hash(&wasm.wasm).to_vec()
-        }
-
         // Gzips a wasm, returning the hash of its compressed representation.
         fn gzip_wasm(wasm: &mut SnsWasm) -> Vec<u8> {
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:5015]: This also seems wrong b/c it's returning a hash instead of a gzipped wasm.  Can you fix this one also?

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 8]
```diff
@@ -1723,6 +1743,56 @@ where
...
+        let all_wasm_hashes = thread_safe_sns.with(|sns_canister| {
+            sns_canister
+                .borrow()
+                .wasm_indexes
+                .keys()
+                .cloned()
+                .collect::<Vec<_>>()
+        });
+        for hash in all_wasm_hashes {
+            ic_cdk_timers::set_timer(std::time::Duration::from_millis(0), move || {
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:1761]: Is it safe to use timers when we haven't migrated from dfn_core?

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 9]
```diff
@@ -1723,6 +1743,56 @@ where
...
             })
     }
+
+    // TODO[NNS1-2997]: Remove this function once the migration is completed.
+    pub async fn populate_wasm_metadata(
+        thread_safe_sns: &'static LocalKey<RefCell<SnsWasmCanister<M>>>,
+    ) {
+        // 2. Compute metadata for all marked WASMs.
+        let all_wasm_hashes = thread_safe_sns.with(|sns_canister| {
+            sns_canister
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:1753]: Since this is going to run on every heartbeat, we probably don't want to create the timers / futures UNLESS there is work to do.  I'd probably do the icheck on the empty metadata here and then produce the list of work to do.  Since it's in a heartbeat, you could just get the next one, if any, that has empty metadata?

Comment (max-dfinity) [reply rs/nns/sns-wasm/src/sns_wasm.rs:1753]: Also, we might add a thread_local static variable to just tell us if there's anything left to do so we can stop doing any work here.  This all seems a little bit expensive perhaps, given there are under 100 wasms, but we'll be checking them all

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 10]
```diff
@@ -338,22 +339,22 @@ where
         &self,
         get_wasm_metadata_payload: GetWasmMetadataRequestPb,
     ) -> Result<Vec<MetadataSection>, String> {
-        let module_hash = <[u8; 32]>::try_from(get_wasm_metadata_payload)?;
+        let hash = <[u8; 32]>::try_from(get_wasm_metadata_payload)?;
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:342]: So the method "get_wasm_metadata" won't use the cached values yet, after this PR is merged?  Should we update that?

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 11]
```diff
@@ -1723,6 +1743,56 @@ where
...
+        });
+        for hash in all_wasm_hashes {
+            ic_cdk_timers::set_timer(std::time::Duration::from_millis(0), move || {
+                let future = async move {
+                    thread_safe_sns.with(|sns_canister| {
+                        if let Some(index) = sns_canister.borrow().wasm_indexes.get(&hash) {
+                            if index.metadata == vec![] {
+                                let metadata = sns_canister
+                                    .borrow()
+                                    .get_wasm_metadata_impl(GetWasmMetadataRequestPb {
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:1768]: Is there a reason to do this as a request?  Instead of using the Self:: read_wasm_metadata_or_err implementation, and then allowing get_wasm_metadata_impl to read from the cache?

**Comment by max-dfinity** [comment]
Maybe dumb question - why not have get_wasm_metadata_impl just read from the cache if it's populated, and then if not, do the work of processing the WASM for the first time?

Then you can avoid all the stuff in heartbeat etc?

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 10 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:342]: Right. The method won't yet be enabled in production. I'm planning to do what you suggest at the time when we enable it (after the migration).

**Comment by aterga** [comment]
> Maybe dumb question - why not have get_wasm_metadata_impl just read from the cache if it's populated, and then if not, do the work of processing the WASM for the first time?
> 
> Then you can avoid all the stuff in heartbeat etc?

I've considered this approach, but that would require making metadata an optional field, and it seems slightly better to me have metadata as non-optional.

**Comment by max-dfinity** [comment]
> I've considered this approach, but that would require making metadata an optional field, and it seems slightly better to me have metadata as non-optional. 

Maybe - is there possibly another approach?  Like tracking which WASMs have been processed?  Then you could remove all that code next upgrade...

File: rs/nervous_system/common/test_utils/src/wasm_helpers.rs (hunk 4 above)
Comment (aterga) [reply rs/nervous_system/common/test_utils/src/wasm_helpers.rs:29]: You're probably referring to the `ensure_sns_wasm_gzipped` function which has the following signature:

```
pub fn ensure_sns_wasm_gzipped(mut sns_wasm: SnsWasm) -> SnsWasm
```

I wanated to have a low-level version of that that doesn't refer to the API type `SnsWasm`. 

As a side remark, ic-wasm that will be used in production for SNS-W uses libflate rather that flate2 (both seem like legit libraries, from a quick google).

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 9 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:1753]: @jasonz-dfinity  pointed out that the whole work should fit into one message, so I'm going to try to make a sync migration and test that. If that works, we won't need any additional state for the migration.

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 8 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:1761]: AFAIK this is safe as we're not mixing IC CDK and dfn_core calls. But we probably won't need it in this PR after all, if we go with a sync migration approach.

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 11 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:1768]: I tried both, this resulted in fewer lines of code.

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 6 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:2209]: Note: This test function was there already before this PR.

I'm going to change it here nonetheless, if I get to it.

File: Cargo.Bazel.Fuzzing.json.lock (hunk 3 above)
Comment (aterga) [reply Cargo.Bazel.Fuzzing.json.lock]: These must be transitive dependencies.

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 9 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs]: It worked, I tested locally using `bazel test --test_env=SSH_AUTH_SOCK //rs/nns/integration_tests:upgrade_canisters_with_golden_nns_state --test_env=NNS_CANISTER_UPGRADE_SEQUENCE=sns-wasm`

**Review by marko-k0 — APPROVED** [review]

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 12]
```diff
@@ -324,36 +325,30 @@ where
...
     ) -> Result<Vec<MetadataSection>, String> {
-        let module_hash = <[u8; 32]>::try_from(get_wasm_metadata_payload)?;
+        let hash = <[u8; 32]>::try_from(get_wasm_metadata_payload)?;
+
+        let Some(wasm) = self.read_wasm(&hash) else {
+            return Err(format!("Cannot find WASM stored under key `{:?}`.", hash));
+        };
 
-        let wasm_metadata = self.read_wasm_metadata_or_err(&module_hash)?;
+        let wasm_metadata = Self::read_wasm_metadata_or_err(&wasm)?;
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:342]: Okay, so we are populating the metadata in the index a post_upgrade hook, but we're still reading it here directly from the wasm?

Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:342]: Oops. Thanks for catching this.

Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs]: Done, PT another L @max-dfinity

File: .github/workflows-source/schedule-daily.yml [hunk 13]
```diff
@@ -132,6 +132,36 @@ jobs:
           ZH2_DLL01_CSV_SECRETS: "${{ secrets.ZH2_DLL01_CSV_SECRETS }}"
           ZH2_FILE_SHARE_KEY: "${{ secrets.ZH2_FILE_SHARE_KEY }}"
 
+  fi-tests-nightly:
```
Comment (max-dfinity) [inline .github/workflows-source/schedule-daily.yml:135]: These seem like they're supposed to be part of a different PR?

Comment (aterga) [reply .github/workflows-source/schedule-daily.yml]: Yes, this was a messy rebase, sorry. I cleaned it up now.

File: rs/nns/sns-wasm/src/sns_wasm.rs [hunk 14]
```diff
@@ -1723,6 +1768,43 @@ where
...
+        for hash in all_wasm_hashes {
+            if let Some(index) = self.wasm_indexes.get(&hash) {
+                if index.metadata != vec![] {
+                    // No need to recompute metadata if it's already set. This code is unlikely to
+                    // actually be executed as this is a synchronous migration.
+                    continue;
+                }
+            }
+            let metadata = self
+                .get_wasm_metadata_impl(GetWasmMetadataRequestPb {
```
Comment (max-dfinity) [inline rs/nns/sns-wasm/src/sns_wasm.rs:1784]: Isn't this supposed to now use read_wasm_metadata_or_err?  B/c get_wasm_metadata_impl now points at the data already in the index?

I think maybe you need to add another test, b/c these probably should have caused a failure somewhere?

Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs:1784]: Ah, yes, you're totally right. This was tested manually, and not the code has evolved in a way that invalidates that test's results. 

Tomorrow, I'll add a test to this branch that would also run on CI (and fix the issue).

File: rs/nns/sns-wasm/BUILD.bazel [hunk 15]
```diff
@@ -32,6 +32,7 @@ BASE_DEPENDENCIES = [
     "@crate_index//:futures",
     "@crate_index//:hex",
     "@crate_index//:ic-cdk",
+    "@crate_index//:ic-cdk-timers",
```
Comment (aterga) [inline rs/nns/sns-wasm/BUILD.bazel:35]:
```suggestion
```

File: rs/nns/sns-wasm/src/sns_wasm.rs (hunk 14 above)
Comment (aterga) [reply rs/nns/sns-wasm/src/sns_wasm.rs]: Fixed and added a test here: https://github.com/dfinity/ic/pull/977/files#diff-2d930ed6199ad023dafa9d3aa47c9c2505fc61ddd028ade4c6b651cd11c253b2R73
